Clarify role mapping support and alternatives (#87572)
elastic/elasticsearch: x-pack/docs/en/security/authorization/mapping-roles.asciidoc

[[mapping-roles]]
=== Mapping users and groups to roles

Role mapping is supported by all realms except `native` and `file`.

> TIP:  Those two realms assign roles directly to users, they do not support role mapping.
> Native realms use <<managing-native-users,user management APIs>>.
> File realms use <<roles-management-file,File-based role management>>.

Reviewer: I'd make this part of the first paragraph, rather than a tip.

```suggestion
The native and file realms assign roles directly to users.
Native realms use <<managing-native-users,user management APIs>>.
File realms use <<roles-management-file,File-based role management>>.
```

> If using Role Mapping, there are two supported sources of role mapping rules,
> <<mapping-roles-api, Role Mapping API>> and <<mapping-roles-file, Role Mapping File>>.

Reviewer: ```suggestion
You can map roles through the 
<<mapping-roles-api, Role mapping API>> (recommended) or a <<mapping-roles-file, Role mapping file>>.
```

> IMPORTANT: Role Mapping API is preferred over Role Mapping File.

Reviewer: ```suggestion
```

> The PKI, LDAP, AD, Kerberos, OpenID Connect, JWT, and SAML realms support
> <<mapping-roles-api, Role Mapping API>>. Only PKI, LDAP, and AD realms
> support <<mapping-roles-file, Role Mapping File>>.

Reviewer: ```suggestion
The PKI, LDAP, AD, Kerberos, OpenID Connect, JWT, and SAML realms support the
<<mapping-roles-api, Role mapping API>>. Only PKI, LDAP, and AD realms
support <<mapping-roles-file, Role mapping files>>.
```

> Besides Role Mapping, the PKI, LDAP, AD, Kerberos, OpenID Connect, JWT, and
> SAML realms realms support <<authorization_realms,Delegated Authorization>> too.
> Role Mapping and Delegated Authorization cannot be used simultaneously in a realm.
> Pick the best option available for each realm based on your requirements.

Reviewer: ```suggestion
The PKI, LDAP, AD, Kerberos, OpenID Connect, JWT, and
SAML realms also support <<authorization_realms,delegated authorization>>.
You can either map roles for a realm or use delegated authorization; you cannot use both simultaneously.
```

> If you choose role mapping, you will have to create roles and _role-mappings_.
> Role mapping rules can based on realm name, realm type, username, groups,
> other user metadata, or combinations of those values.

Reviewer: ```suggestion
To use role mapping, you create roles and role mapping rules.
Role mapping rules can based on realm name, realm type, username, groups,
other user metadata, or combinations of those values.
```

NOTE: When <<anonymous-access,anonymous access>> is enabled, the roles
of the anonymous user are assigned to all the other users as well.

> There are two sources of role-mapping rules.

Reviewer: I'd replace this paragraph with the suggestion below. Can't suggest changes to non-modified text. :-)

```suggestion
If there are role-mapping rules created through the API as well as a role mapping file, 
the rules are combined. 
It's possible for a single user to have some roles that were mapped through the API, 
and others assigned based on the role mapping file.
``````

the API, and other roles that are mapped through files.

NOTE: Users with no roles assigned will be unauthorized for any action.
In other words, they may be able to authenticate, but they will have no roles.
No roles means no privileges, and no privileges means no authorizations to
make requests.

> When you use role-mappings to assign roles to users, the roles must exist.

Reviewer: ```suggestion
When you use role mappings to assign roles to users, the roles must exist.
```
